ftp/pfSense-pkg-tftpd/files/usr/local/www/tftp_files.php

```diff
@@ -0,0 +1,230 @@
+	header("Cache-Control: no-cache, must-revalidate");
+	header("Expires: Sat, 26 Jul 1997 05:00:00 GMT");
+	header("Content-Length: " . filesize("{$desc}"));
+	fpassthru($fd);
+	exit;
+}
+
+/* Restore TFTP server backup */
+if ($_GET['a'] == "other" && $_GET['t'] == "restore") {
+		if (file_exists($backup_path)) {
+			conf_mount_rw();
+			mwexec("/usr/bin/tar -xpzC / -f {$backup_path}");
+			header("Location: tftp_files.php?savemsg=Backup+has+been+restored.");
+			conf_mount_ro();
+		} else {
+			header("Location: tftp_files.php?savemsg=Restore+failed.+Backup+file+not+found.&result=alert-warning");
+		}
+		exit;
+}
+
+/* Upload files to TFTP server */
+if ($_POST['upload'] == "Upload" && $_FILES["tftpd_fileup"]["error"] == UPLOAD_ERR_OK) {
+	if (is_uploaded_file($_FILES['tftpd_fileup']['tmp_name'])) {
+		conf_mount_rw();
+		$tmp_name = $_FILES["tftpd_fileup"]["tmp_name"];
+		$name = basename($_FILES["tftpd_fileup"]["name"]);
+		move_uploaded_file($tmp_name, "{$files_dir}/{$name}");
+		conf_mount_ro();
+	} else {
+		$input_errors[] = gettext("Failed to upload file {$_FILES["tftpd_fileup"]["name"]}");
+	}
+}
+
+/* Delete a file from TFTP server */
+if ($_GET['act'] == "del") {
+	if ($_GET['type'] == 'tftp') {
+		conf_mount_rw();
+		$filename = htmlspecialchars($_GET['filename']);
+		// Only delete files under the $files_dir!
+		$basedirlength = strlen($files_dir);
+		if (substr($filename, 0, $basedirlength) !== "{$files_dir}") {
+			$error_msg = "Attempt to delete files outside of TFTP server directory rejected!";
+			log_error("[tftpd] {$error_msg}");
+			file_notice("tftpd", $error_msg, "Packages");
+			header("Location: tftp_files.php");
+			return;
+		} else {
+			unlink_if_exists("{$filename}");
+			conf_mount_ro();
+			header("Location: tftp_files.php");
+			exit;
+		}
+	}
+}
+
+$pgtitle = array("Package", "TFTP server", "Files");
+require_once("head.inc");
+$savemsg = htmlspecialchars($_GET["savemsg"]);
+$result = htmlspecialchars($_GET["result"]) ?: 'success';
+if ($savemsg) {
+	print_info_box($savemsg, $result);
+}
+$tab_array = array();
+$tab_array[] = array("Settings", false, "/pkg_edit.php?xml=tftpd.xml&amp;id=0");
+$tab_array[] = array("Files", true, "/tftp_files.php");
+display_top_tabs($tab_array);
+?>
+
+<div class="panel panel-default">
+	<div class="panel-heading"><h2 class="panel-title"><?=gettext("TFTP Server Files Management")?></h2></div>
+	<div class="panel-body table-responsive">
+		<table class="table table-striped table-hover table-condensed">
+		<tbody><tr><td>
+			<table class="table table-striped table-hover table-condensed">
+			<thead>
+				<th>File Name</th>
+				<th>Last Modified</th>
+				<th>Size</th>
+				<th>Actions</th>
+			</thead>
+			<tbody>
+			<?php $tftpdfiles = new RecursiveDirectoryIterator("{$files_dir}"); ?>
+			<?php foreach (new RecursiveIteratorIterator($tftpdfiles) as $filename => $file): ?>
+			<?php if (is_file($file)): ?>
+				<tr>
+					<td><?=gettext($file); ?></td>
```

Comment:
Why `gettext()` on a filename? Shouldn't we display that exactly as it is?
Also, probably a good idea to use `htmlspecialchars()` before displaying that filename to the browser.